File: Flink/src/test/java/com/lancer/scala/Test03.scala
/*
package com.lancer.scala

import com.lancer.scala.model.Person
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.api.common.typeinfo.{TypeHint, TypeInformation}
import org.apache.flink.api.scala.typeutils.Types
import org.apache.flink.streaming.api.datastream.DataStreamSink
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.scala.{DataStream, OutputTag, StreamExecutionEnvironment}
import org.apache.flink.streaming.connectors.kafka.{FlinkKafkaConsumer, FlinkKafkaProducer, KafkaDeserializationSchema, KafkaSerializationSchema}
import org.apache.flink.util.Collector
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.producer.ProducerRecord

import java.lang
import java.nio.charset.StandardCharsets
import java.util.Properties

object Test03 {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    val consumer = new Properties()
    val producer = new Properties()

    consumer.load(this.getClass.getClassLoader.getResourceAsStream("consumer.properties"))
    producer.load(this.getClass.getClassLoader.getResourceAsStream("producer.properties"))

    import org.apache.flink.api.scala._

    // 如果只需要value
    val source: DataStream[String] = env.addSource[String](new FlinkKafkaConsumer[String]("flinkConsole", new SimpleStringSchema(), consumer))

    /* // 如果需要对kafka里面的kv进行操作
    val source = env.addSource[String](new FlinkKafkaConsumer[String]("flinkConsole", new KafkaDeserializationSchema[String] {
      override def isEndOfStream(t: String): Boolean = false

      override def deserialize(consumerRecord: ConsumerRecord[Array[Byte], Array[Byte]]): String = {
        new String(consumerRecord.value(), StandardCharsets.UTF_8)
      }

      override def getProducedType: TypeInformation[String] = {
        // createTypeInformation[String]
        // TypeInformation.of(classOf[String])
        Types.STRING
      }
    }, consumer))*/

    // 解决使用泛型嵌套中，遇到的问题
    // TypeInformation.of(classOf[(String, Int)])
    // TypeInformation.of(new TypeHint[(String, Int)] {})
    // TypeInformation.of(new TypeHint[((String, Int), (String, (Int, String)))] {})

    val outputTag = OutputTag[String]("adult")

    val outputDataStream: DataStream[Person] = source
      .filter(
        (line: String) => {
          if (!"".equals(line)) {
            val splits = line.split(",")
            splits.length == 3 && (splits(2).trim.toInt >= 0 && splits(2).trim.toInt <= 100)
          } else {
            false
          }
        })
      .process[Person](
        new ProcessFunction[String, Person]() {
          override def processElement(value: String, ctx: ProcessFunction[String, Person]#Context, out: Collector[Person]): Unit = {
            val splits = value.split(",");
            val name = splits(0).trim();
            val gender = splits(1).trim();
            val age = Integer.parseInt(splits(2).trim());
            val person = new Person(name, gender, age);
            if (age >= 18) {
              out.collect(person);
            } else {
              ctx.output(outputTag, value);
            }
          }
        }
      )

    outputDataStream
      .print()

    val resultStream: DataStreamSink[String] = outputDataStream
      .getSideOutput[String](outputTag)
      .addSink(new FlinkKafkaProducer[String]("flinkTest", new MyKafkaSerializationSchema, producer, FlinkKafkaProducer.Semantic.EXACTLY_ONCE)
      )
    env.execute(this.getClass.getSimpleName)
  }

  class MyKafkaSerializationSchema extends KafkaSerializationSchema[String] {
    override def serialize(element: String, timestamp: lang.Long): ProducerRecord[Array[Byte], Array[Byte]] = new ProducerRecord("flinkTest", element.getBytes(StandardCharsets.UTF_8))
  }
}
*/

File: Flink/src/test/java/com/lancer/java/Test03.java
package com.lancer.java;

import com.lancer.java.model.Person;
import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer;
import org.apache.flink.streaming.connectors.kafka.KafkaSerializationSchema;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;
import org.apache.kafka.clients.producer.ProducerRecord;

import javax.annotation.Nullable;
import java.nio.charset.StandardCharsets;
import java.util.Properties;

public class Test03 {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        Properties consumer = new Properties();
        Properties producer = new Properties();

        consumer.load(Test.class.getClassLoader().getResourceAsStream("consumer.properties"));
        producer.load(Test.class.getClassLoader().getResourceAsStream("producer.properties"));

        OutputTag<String> outputTag = new OutputTag<String>("adult") {
        };

        DataStreamSource<String> source = env.addSource(new FlinkKafkaConsumer<>("flinkConsole", new SimpleStringSchema(), consumer));

        SingleOutputStreamOperator<Person> outputDataStream = source
                .filter(new FilterFunction<String>() {
                    @Override
                    public boolean filter(String value) throws Exception {
                        if (!"".equals(value)) {
                            String[] splits = value.split(",");
                            return splits.length == 3 && (Integer.parseInt(splits[2]) >= 0 && Integer.parseInt(splits[2]) <= 100);
                        }
                        return false;
                    }
                })
                // 判断是否成年，成年流直接打印，未成年流放入kafka
                .process(new ProcessFunction<String, Person>() {
                    @Override
                    public void processElement(String value, Context ctx, Collector<Person> out) throws Exception {
                        String[] splits = value.split(",");
                        String name = splits[0].trim();
                        String gender = splits[1].trim();
                        int age = Integer.parseInt(splits[2].trim());
                        Person person = new Person(name, gender, age);
                        if (age >= 18) {
                            out.collect(person);
                        } else {
                            ctx.output(outputTag, value);
                        }
                    }
                });
        outputDataStream.print();

        outputDataStream
                .getSideOutput(outputTag)
                .addSink(new FlinkKafkaProducer<String>("flinkTest", (KafkaSerializationSchema<String>) (element, timestamp) -> new ProducerRecord<>("flinkTest", element.getBytes(StandardCharsets.UTF_8)), producer, FlinkKafkaProducer.Semantic.EXACTLY_ONCE));

        env.execute(Test.class.getSimpleName());
    }
}
